fsthttp/imageopto/imageopto.go

```diff
@@ -0,0 +1,964 @@
+//
+// Valid values are either an integer number of pixels, or a percentage followed by `p`.
+type PixelsOrPercentage string
+
+func (p PixelsOrPercentage) isSet() bool    { return p != "" }
+func (p PixelsOrPercentage) String() string { return string(p) }
+
+func (p PixelsOrPercentage) validate() error {
+	if !rePixelsOrPercentage.MatchString(string(p)) {
+		return fmt.Errorf("imageopto: pixels or percentage: invalid: %q", string(p))
+	}
+
+	return nil
+}
+
+var rePixelsOrPercentage = regexp.MustCompile("^[0-9]+(.[0-9]+)?p?$")
+
+// CropMode determines cropping behavior.
+type CropMode string
+
+const (
+	// CropModeSafe avoids invalid parameter errors and return an image.
+	//
+	// By default, if the specified cropped region is outside the bounds of the image, then the
+	// transformation will fail with the error "Invalid transformation for requested image:
+	// Invalid crop, region out of bounds". This option will instead deliver the image as an
+	// intersection of the origin image and the specified cropped region. This avoids the error,
+	// but the resulting image may not be of the specified dimensions.
+	CropModeSafe = "safe"
+
+	// CropModeSmart enables content-aware algorithms to attempt to crop the image to the desired aspect ratio
+	// while intelligently focusing on the most important visual content, including the detection
+	// of faces.
+	CropModeSmart = "smart"
+)
+
+func (c CropMode) isSet() bool    { return c != "" }
+func (c CropMode) String() string { return string(c) }
+
+type areaState uint8
+
+const (
+	areaStateNone areaState = iota
+	areaStateAspectRatio
+	areaStateWidthHeight
+)
+
+type areaAspectRatio struct{ w, h uint32 }
+type areaWidthHeight struct{ w, h PixelsOrPercentage }
+
+// Area is an image area.
+type Area struct {
+	aspect      areaAspectRatio
+	widthHeight areaWidthHeight
+	state       areaState
+}
+
+// NewAreaAspectRatio specifies an area with the given width/height aspect ratio.
+func NewAreaAspectRatio(w, h uint32) Area {
+	return Area{
+		aspect: areaAspectRatio{w, h},
+		state:  areaStateAspectRatio,
+	}
+}
+
+// NewAreaWidthHeight specifies an area with precise number of pixels.
+func NewAreaWidthHeight(w, h PixelsOrPercentage) Area {
+	return Area{
+		widthHeight: areaWidthHeight{w, h},
+		state:       areaStateWidthHeight,
+	}
+}
+
+func (a *Area) validate() error {
+	if a.state == areaStateWidthHeight {
+		if a.widthHeight.w.isSet() {
+			if err := a.widthHeight.w.validate(); err != nil {
+				return err
+			}
+		}
+		if a.widthHeight.h.isSet() {
+			if err := a.widthHeight.h.validate(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (area *Area) String() string {
+	switch area.state {
+	case areaStateNone:
+		return ""
+	case areaStateAspectRatio:
+		return fmt.Sprintf("%v:%v", area.aspect.w, area.aspect.h)
+	case areaStateWidthHeight:
+		return fmt.Sprintf("%v,%v", area.widthHeight.w.String(), area.widthHeight.h.String())
```

Comment:
I think you could change all of these to use the `Stringer` implicitly, even leaving the `%v` (vs. changing to `%s`), e.g.

```suggestion
		return fmt.Sprintf("%v,%v", area.widthHeight.w, area.widthHeight.h)
```

Reply:
I think I'll merge as is and we can remove `fmt` later.  I'll file an issue.